### PR TITLE
Allow to control LIBC_WASI from command line options

### PR DESCRIPTION
**Pull request by lum1n0us**
$ cmake -S . -B build -DWAMR_BUILD_LIBC_WASI=0

plus, add several messges to show options status

### CODE REVIEW

**Review by yamt — COMMENTED** [review]
how about `add_definitions(-DWASM_ENABLE_LIB_WASI_THREADS=1)` around L47 in the file?

i guess it's nicer to make them runtime options.

**Comment by lum1n0us** [comment]
we will use `WAMR_BUILD_LIB_WASI_THREADS` to control `include (lib_wasi_threads.cmake)`, and to set `WASM_ENABLE_LIB_WASI_THREADS`. By default, `WAMR_BUILD_LIB_WASI_THREADS` is 1.

YES. `option` is better. but it is not WAMR style so far.

**Comment by yamt** [comment]
> YES. `option` is better. but it is not WAMR style so far.

what's "WAMR style"? suitable for embedded?
as wamr-compiler involves LLVM, i'm not sure if it makes much sense to follow the same "embedded" style.

also, we already have an option to control similar things. (--native-lib)

i don't insist as it would be more involved changes than this PR though.

File: wamr-compiler/CMakeLists.txt [hunk 1]
```diff
@@ -198,32 +197,82 @@ include_directories (${SHARED_DIR}/include
...
+if (WAMR_BUILD_LIBC_BUILTIN EQUAL 1)
+  include (${IWASM_DIR}/libraries/libc-builtin/libc_builtin.cmake)
+endif()
+
+if (WAMR_BUILD_LIBC_WASI EQUAL 1)
+  if (NOT MINGW)
+    if (NOT MSVC)
+      include (${IWASM_DIR}/libraries/libc-wasi/libc_wasi.cmake)
+    else()
+      include (${IWASM_DIR}/libraries/libc-uvwasi/libc_uvwasi.cmake)
```
Comment (yamt) [inline wamr-compiler/CMakeLists.txt:263]: i feel it's better to use WAMR_BUILD_LIBC_UVWASI to control libc_uvwasi to align better with config_common.cmake.

Comment (wenyongh) [reply wamr-compiler/CMakeLists.txt]: So, it should be like below:
```cmake
if (NOT MINGW AND NOT MSVC)
  if (NOT DEFINED WAMR_BUILD_LIBC_WASI)
    set(WAMR_BUILD_LIBC_WASI 1)
  endif()
else ()
  if (NOT DEFINED WAMR_BUILD_LIBC_UVWASI)
    set(WAMR_BUILD_LIBC_UVWASI 1)
  endif()
endif()
```
yes, seems it is better.

File: wamr-compiler/CMakeLists.txt [hunk 2]
```diff
@@ -199,31 +198,98 @@ include_directories (${SHARED_DIR}/include
...
-else()
-    set(WAMR_BUILD_LIBC_UVWASI 1)
-endif()
+  if ((NOT DEFINED WAMR_BUILD_LIBC_WASI) AND (NOT DEFINED WAMR_BUILD_LIBC_UVWASI))
+    message ("-- enable WAMR_BUILD_LIBC_WASI by default")
+    set (WAMR_BUILD_LIBC_WASI 1)
+  endif ()
+
+  if ((WAMR_BUILD_LIBC_WASI EQUAL 1) AND (WAMR_BUILD_LIBC_UVWASI EQUAL 1))
+    message (WARNING "-- pick WAMR_BULID_LIBC_UVWASI when both are enable")
```
Comment (wenyongh) [inline wamr-compiler/CMakeLists.txt:207]: `are enabled`

File: wamr-compiler/CMakeLists.txt [hunk 3]
```diff
@@ -199,31 +198,98 @@ include_directories (${SHARED_DIR}/include
 enable_language (ASM)
 
 if (NOT MINGW AND NOT MSVC)
-    set(WAMR_BUILD_LIBC_WASI 1)
-else()
-    set(WAMR_BUILD_LIBC_UVWASI 1)
-endif()
+  if ((NOT DEFINED WAMR_BUILD_LIBC_WASI) AND (NOT DEFINED WAMR_BUILD_LIBC_UVWASI))
+    message ("-- enable WAMR_BUILD_LIBC_WASI by default")
```
Comment (wenyongh) [inline wamr-compiler/CMakeLists.txt:202]: Should we add `WARNING`? There will be following output `-- Libc WASI enabled`

File: wamr-compiler/CMakeLists.txt [hunk 4]
```diff
@@ -199,31 +198,98 @@ include_directories (${SHARED_DIR}/include
...
+if (WAMR_BUILD_LIBC_UVWASI EQUAL 1)
+  message ("-- Libc WASI enabled with uvwasi implementation")
+endif ()
+
+if (WAMR_BUILD_LIBC_WASI EQUAL 1)
+  message ("-- Libc WASI enabled")
+endif ()
+
+if ((NOT WAMR_BUILD_LIBC_WASI) AND (NOT WAMR_BUILD_LIBC_UVWASI))
+  message ("-- Libc WASI disable")
```
Comment (wenyongh) [inline wamr-compiler/CMakeLists.txt:242]: disabled

File: wamr-compiler/CMakeLists.txt [hunk 5]
```diff
@@ -199,31 +198,98 @@ include_directories (${SHARED_DIR}/include
...
+endif ()
+
+if (WAMR_BUILD_LIBC_BUILTIN EQUAL 1)
+  message ("-- Libc builtin enabled")
+else ()
+  message ("-- Libc builtin disable")
```
Comment (wenyongh) [inline wamr-compiler/CMakeLists.txt:248]: disabled

File: wamr-compiler/CMakeLists.txt [hunk 6]
```diff
@@ -199,31 +198,98 @@ include_directories (${SHARED_DIR}/include
...
+endif ()
+
+if (WAMR_BUILD_LIB_PTHREAD EQUAL 1)
+  message ("-- Lib pthread enabled")
+else ()
+  message ("-- Lib pthread disable")
```
Comment (wenyongh) [inline wamr-compiler/CMakeLists.txt:254]: disabled

File: wamr-compiler/CMakeLists.txt [hunk 7]
```diff
@@ -199,31 +198,98 @@ include_directories (${SHARED_DIR}/include
...
+endif ()
+
+if (WAMR_BUILD_LIB_WASI_THREADS EQUAL 1)
+  message ("-- Lib wasi-threads enabled")
+else ()
+  message ("-- Lib wasi-threads disable")
```
Comment (wenyongh) [inline wamr-compiler/CMakeLists.txt:260]: disabled

**Review by wenyongh — APPROVED** [review]
LGTM

**Pull request merged**